### PR TITLE
Assembler now exits with an error code upon encountering a fatal error

### PR DESCRIPTION
**Pull request by C-Elegans**
This has been added because not exiting when an error has been encountered
has made it much more difficult to track find bugs in my projects. Instead
of building the project (I've been using a makefile, and the error output
gets hidden most of the time) and wondering why my program will not work,
the build will immediately fail, with the error right at the bottom of the
terminal.

### CODE REVIEW

File: errors.cpp [hunk 1]
```diff
@@ -152,6 +152,7 @@ static void ReplayErrorRecursive(const list_t *pList, bool fFatalOnly)
 		}
 		PrintSPASMError(lpError);
 	}
+	exit(-1);
```
Comment (tari) [inline errors.cpp]: This should probably be a positive nonzero value, to avoid confusion with signal termination on unixes.

**Comment by tari** [comment]
I don't think inside error printing is a good place to be exiting the program on error. I'd rather make `EndSPASMErrorSession` return something like the number of fatal and nonfatal errors, then the call sites in `main` can exit with an appropriate status if there are problems.

**Comment by tari** [comment]
(This looks like a fix for #33.)

**Comment by alberthdev** [comment]
Yup, agreed - considering that `ReplayErrorRecursive` is called recursively, this would result in only printing one error, instead of all of the errors. (`exit` would be called before it returned, ending the program.) A quick test shows this behavior:

    # Created test.z80 with junk:
    #     af
    #     geiwjoigi
    #     f f
    #     29r30tu4)G$(
    
    $ ./spasm test.z80
    Pass one... 
    test.z80:3: error SE113: Unknown opcode 'f'
    test.z80:4: error SE119: Unknown syntax
    Pass two... 
    Assembly time: 0.024 seconds
    
    # Added the exit(1) here to errors.cpp, then rebuilt:
    
    $ make
    g++ -DUSE_REUSABLES -DUNIXVER -DUSE_BUILTIN_FCREATE -c errors.cpp
    g++ -o spasm main.o opcodes.o pass_one.o pass_two.o utils.o export.o preop.o directive.o console.o expand_buf.o hash.o list.o parser.o storage.o errors.o bitmap.o modp_ascii.o opcodes_ez80.o -lc -lm -lgmp -lcrypto
    
    $ ./spasm test.z80
    Pass one... 
    test.z80:3: error SE113: Unknown opcode 'f'
    
    $

As for a potential fix, something handled in `main()` is possible... though I'm thinking of making it "nice" and aiming for something like `EndSPASMErrorSession(CleanupSPASMErrorSession(...))`, where the functionality of the current `EndSPASMErrorSession` is moved to `CleanupSPASMErrorSession`, the new `CleanupSPASMErrorSession` returns the error statistics (no longer `void`), and the new `EndSPASMErrorSession` handles the exit logic based on those stats.

That way, we can keep the error handling logic within `errors.cpp`, and still decide whether to exit or not within `main`.

Any thoughts on all of this?

**Comment by alberthdev** [comment]
Thanks for the update! Just curious if you could try to aim for the `EndSPASMErrorSession(CleanupSPASMErrorSession(...))` API I proposed, if it's possible! Also, please fix the indentation/line breaks to be more consistent with the source.

**Comment by C-Elegans** [comment]
yeah, sorry about that, they show up fine in the editor. The reason I didn't go for the EndSPASMErrorSession(CleanupSPASMErrorSession(...)) is because it would require changing every call to EndSPASMErrorSession, but I'll try changing it

File: main.cpp [hunk 2]
```diff
@@ -129,22 +130,26 @@ int run_assembly()
 	
 	list_free (include_dirs, true, NULL);
 	include_dirs = NULL;
-
+    if(firstPassFatalErrors >0){
+        exit(firstPassFatalErrors); //or should this be just a constant?
```
Comment (tari) [inline main.cpp]: This should be a constant. A return code greater than 255 is not very portable, so something like 256 errors might be reported back to a calling process as code 0 due to truncation.

**Comment by alberthdev** [comment]
Thanks for the fix - merged!

**Pull request merged**